### PR TITLE
createAction does more pre-processing

### PR DESCRIPTION
**Pull request by Lucretiel**
Slightly simplified createAction, and made it do more processing ahead of time that doesn't have to be done every time an action is emitted

### CODE REVIEW

File: src/createAction.js [hunk 1]
```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
     'Expected payloadCreator to be a function, undefined or null'
   );
 
-  const finalPayloadCreator = isNull(payloadCreator)
+  const finalPayloadCreator = payloadCreator === null || payloadCreator === identity
```
Comment (yangmillstheory) [inline src/createAction.js]: Not sure why we're not using `isNull` anymore? Also why is the `identity` check needed?

File: src/createAction.js [hunk 2]
```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
...
+    
+    if (hasMeta) {
       action.meta = metaCreator(...args);
     }
 
     return action;
   };
 
-  actionCreator.toString = () => type.toString();
+  const typeStr = type.toString();
```
Comment (yangmillstheory) [inline src/createAction.js]: Is the local declaration needed, given that it's used in one place?

File: src/createAction.js [hunk 3]
```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
...
-    const hasError = args[0] instanceof Error;
-
-    const action = {
-      type
-    };
-
-    const payload = hasError ? args[0] : finalPayloadCreator(...args);
-    if (!isUndefined(payload)) {
+    const payload = finalPayloadCreator(...args);
+    const action = { type, error: payload instanceof Error };
```
Comment (yangmillstheory) [inline src/createAction.js]: This makes the key `error` always be present. Do we really want this new behavior?

File: src/createAction.js [hunk 4]
```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
...
+    
+    if (payload !== undefined) {
```
Comment (yangmillstheory) [inline src/createAction.js]: Wait what's the objection to using `isUndefined`? Does it cost that much bytes that the readability isn't worth it? It's not much less tokens to parse, but still.

File: src/createAction.js [hunk 5]
```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
...
     }
-
-    if (hasError || payload instanceof Error) {
-      // Handle FSA errors where the payload is an Error object. Set error.
-      action.error = true;
-    }
-
-    if (isFunction(metaCreator)) {
+    
+    if (hasMeta) {
```
Comment (yangmillstheory) [inline src/createAction.js]: So this is where the savings is?

File: src/createAction.js (hunk 1 above)
Comment (Lucretiel) [inline src/createAction.js]: 1. No reason. I just figured that the purpose of `isNull` is to be a predicate to functions like `filter`, so there's no reason to have the function call overhead here.
2. The `identity` check means that if the `payloadCreator` is `identity`, we can skip the `instanceof Error` check every time it's called, and don't have to use the latter version of the `finalPayloadCreator`.

File: src/createAction.js (hunk 3 above)
Comment (Lucretiel) [inline src/createAction.js]: I don't have a strong preference either way. It doesn't change anything for FSA-compliant applications, and I liked the simplicity of this version. I would have put `meta` and `payload` inline as well, but there was no good way to do that given the limitations of javascript syntax.

File: src/createAction.js (hunk 4 above)
Comment (Lucretiel) [inline src/createAction.js]: It's not a readability thing; just avoiding the function call overhead. I figured that the purpose of `isUndefined` is to be used as a predicate, like in `filter`, but there's no reason to use it here, when the comparison is so simple. Compare to `isFunction`, which has a much more complicated definition.

File: src/createAction.js (hunk 2 above)
Comment (Lucretiel) [inline src/createAction.js]: It means that we don't have to call `type.toString` every time `actionCreator.toString` is called.

File: src/createAction.js (hunk 5 above)
Comment (Lucretiel) [inline src/createAction.js]: Yes, especially given that `isFunction` is surprisingly involved.

**Comment by Lucretiel** [comment]
Obviously, all of this stuff is based on my personal preference. Originally I was only planning on doing the `hasMeta` change, but then I started editing other stuff while I was here. I've given my own justification for all these things, but feel free to give me a firm yes or no on any of them.

**Comment by yangmillstheory** [comment]
So, these are actually quite good reasons. The build is failing because of lint; can you get that to pass and I'll get this merged Thanks.

File: src/createAction.js (hunk 1 above)
Comment (yangmillstheory) [inline src/createAction.js]: 1. re: overhead, this is called just once during when defining the action creator, so do we need to worry about it? If not, I would advise removing the `isFunction` call above and below as well for consistency, but would prefer to just keep as is.

File: src/createAction.js (hunk 4 above)
Comment (yangmillstheory) [inline src/createAction.js]: Makes sense here, sense this is being called dynamically at runtime.

File: src/createAction.js (hunk 2 above)
Comment (yangmillstheory) [inline src/createAction.js]: Can we move this declaration up a scope?

**Comment by yangmillstheory** [comment]
@Lucretiel I fixed some lint issues in the PR. I would like to get this in, but some tests are failing. Could you fix?

**Comment by Lucretiel** [comment]
Yeah, mostly it has to do with the having `error` always be present. I'll just remove that aspect, since I didn't feel that strongly about it anyway.

**Comment by Lucretiel** [comment]
Do you mind if I do a force push? Or did you make changes you feel strongly about?

**Comment by yangmillstheory** [comment]
Looks good now. Can you take a look and tell me if this is OK with you? I'll merge and publish a patch version.

**Comment by Lucretiel** [comment]
Made one final change. Everything should be good now.

**Comment by yangmillstheory** [comment]
Thanks 👍 .

**Comment by yangmillstheory** [comment]
https://github.com/acdlite/redux-actions/releases/tag/v2.0.3

```
~/c/p/redux-actions (45c3a78)|master✓
$ npm search redux-actions
NAME                      | DESCRIPTION          | AUTHOR          | DATE       | VERSION  | KEYWORDS
redux-actions             | Flux Standard…       | =acdlite…       | 2017-05-10 | 2.0.3    |
redux-actions-assertions- | Javascript…          | =dmitry-zaets   | 2016-09-12 | 1.1.0    |
js                        |                      |                 |            |          |
redux-actions-sequences   | Make sequences of…   | =andrevinsky    | 2017-01-17 | 1.1.19   |
npmdoc-redux-actions      | #### api…            | =npmdoc         | 2017-04-24 | 2017.4.… |
redux-actions-helpers     | Small helpers…       | =olegman        | 2016-10-27 | 1.0.2    |
redux-actions-api-addon   | Redux Actions…       | =jkusachi       | 2016-07-14 | 1.1.0    |
redux-actions-utils       | Utility fo Redux…    | =dahel          | 2017-03-27 | 0.0.2    |
xm-redux-actions          | Flux Standard…       | =yeanzhi        | 2017-04-24 | 2.0.2    |
redux-actions-assertions- | Javascript…          | =dmitry-zaets   | 2016-09-12 | 1.1.0    |
js                        |                      |                 |            |          |
redux-actions-sequences   | Make sequences of…   | =andrevinsky    | 2017-01-17 | 1.1.19   |
npmdoc-redux-actions      | #### api…            | =npmdoc         | 2017-04-24 | 2017.4.… |
redux-actions-helpers     | Small helpers…       | =olegman        | 2016-10-27 | 1.0.2    |
redux-actions-api-addon   | Redux Actions…       | =jkusachi       | 2016-07-14 | 1.1.0    |
redux-actions-utils       | Utility fo Redux…    | =dahel          | 2017-03-27 | 0.0.2    |
```

**Pull request merged**